api/v1alpha1/authorization_types.go

```diff
@@ -47,17 +50,59 @@ type Principal struct {
 	// ClientCIDRs are the IP CIDR ranges of the client.
 	// Valid examples are "192.168.1.0/24" or "2001:db8::/64"
 	//
+	// If multiple CIDR ranges are specified, one of the CIDR ranges must match
+	// the client IP for the rule to match.
+	//
 	// The client IP is inferred from the X-Forwarded-For header, a custom header,
 	// or the proxy protocol.
 	// You can use the `ClientIPDetection` or the `EnableProxyProtocol` field in
 	// the `ClientTrafficPolicy` to configure how the client IP is detected.
 	// +kubebuilder:validation:MinItems=1
 	ClientCIDRs []CIDR `json:"clientCIDRs"`
 
-	// TODO: Zhaohuabing the MinItems=1 validation can be relaxed to allow empty list
-	// after other principal types are supported. However, at least one principal is required
+	// JWTClaims are the claims in a JWT token.
+	//
+	// If multiple claims are specified, all claims must match for the rule to match.
+	// For example, if there are two claims: one for the audience and one for the issuer,
+	// the rule will match only if both the audience and the issuer match.
+	// Note: in order to use JWT claims for authorization, you must configure the
+	// JWT authentication in the same `SecurityPolicy`.
+	// +optional
+	// +notImplementedHide
+	JWTClaims []JWTClaim `json:"jwtClaims"`
 }
 
+// JWTClaim specifies a claim in a JWT token.
+type JWTClaim struct {
```

Comment:
thoughts on this ?
```
- type: Exact
  name: foo
  value: bar
- type: Contains
  name: foo
  value: baz
```

this API allows for exact, contains and can even be extended for regex if needed, and `exact` can be default

Reply:
`type` / `matchType`

Reply:
Is`Contains` matchType for string, string array, or both? If it works for both, it can be confusing.

Reply:
Thought it again. If we convert bool and int to string, a `ClaimType` here is misleading and a `MatchType` makes more sense.

So my understanding is:

- `Exact` is used to match a string/bool/int claim type.
- `Contains` is used to match a string Array claim type.

`Regex` can be confusing, better not to support it.

Reply:
Bit sure that I see the value of `Contains` or regex for the name of the claim. Would be helpful for the values. If forced to choose between `Contains` and the array for the values though, I think they are both valid use cases but do not completely overlap. If I want to allow a specific claim name but with one of x number of known values, then the array specifier would work for that, but `Contains` might not (depending on the values). But for example, if there were one where the value just needed to contain something, that could definitely be useful over the array of exacts since you might not be capable of easily specifying all accepted values with that method.

Reply:
hey @AliceProxy this is the example we are trying to solve with `Contains` , were you thinking of the same case ?

```
- matchType: Contains
  name: foo
  value: baz
```  

```
   "foo": [
      "bar",
      "baz"
   ]
```

another example is the `amr` claim highlighted in https://developer.okta.com/docs/guides/customize-tokens-returned-from-okta/main/#request-a-token-with-the-custom-claim

Reply:
@arkodg That makes sense. Ilike I mentioned, I think it is useful for the value. Maybe it's just me, but I see the name directly under matchType and it is unclear to me that it would only apply to the value since they are all on the same indent/object level.

Reply:
maybe just changing it to `valueType` would add clarity? I still do think there are cases where I would still want the slice instead of `Contains` if they don't share common substrings, etc. and we don't plan on supporting `Regex` as a type.

Reply:
I personally don't have any issues with Regex. I am quite fond of it for certain use-cases.

Reply:
adding some more links to similar work in other libraries/projects
* golang `slices` has a `Contains` API https://pkg.go.dev/golang.org/x/exp/slices#Contains, but since its a method within `Slices`, it more explanatory than matchType
* CEL https://github.com/google/cel-spec/blob/v0.7.0/doc/langdef.md#macros has `has` and `exists`
* the sets package has `Has` https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets#String.Has

Reply:
The term “MatchType” seems a bit vague to me - it implicitly suggests the type of the claim value(string vs string array) without being clear. I would prefer a more explicit name, such as "ValueType" or "ClaimType".

Given that bool and int values are converted to string, maybe "ValueType" is the better naming.

Reply:
another reason why I proposed the alternate API is
* the `values` field is a list, which made it unclear to what how the match was happening
* the `String` and `StringArray` fields tell me the data type of the `claim` but doesnt tell me how I'm matching on them

Reply:
Unlike routing or rate limiting, which may use various matching types, "Exact“ is typically the sole method used for matching the principals(users, roles, groups, etc). Other kinds of matching types such as "Prefix", "Regex", and "Contains" are rarely used and might seem unusual if supported. Thus, 'Exact` aligns users' expectations.

In the context of `StringArray`, the specified value must "exactly" match one of the elements in the claim value, so it also seems clear to me.

Reply:
The scope claim is "special" because it isn't an array, is a space delimited list. 

When you want to check if a scope is present, it is very much a "contains".

Imagine you have a route/resource that requires the scope books:read however the claim scope will include more scopes in the string and they're space delimited.

@zhaohuabing how are we enabling us to check the presence of that a specific scope like books:read is present in the space delimited list in the scopes claim?

I've had to use regex in the past in my life and I really dislike it, especially since you don't know if the scope is at the beginning middle or end of the space delimited list.

Reply:
@missBerg The authorization is intended to be used against ID tokens, where the scopes should already be resolved to fine-grained claims (The second column in the following table).


Scope | Claims
-- | --
email | email, email_verified
address | address
profile | name, family_name, given_name, middle_name, nickname, preferred_username, profile, picture, website, gender, birthdate, zoneinfo, locale, updated_at
phone | phone_number, phone_number_verified
openid | sub, auth_time, acr

Reply:
Do we need a different issue to enable AuthZ for JWT access tokens? 

It's a super valuable and important feature.

Evaluating the authorization level of an access token is the most typical token to use when deciding if a request should be routed onto the underlying target resource.

ID tokens are not proofs of authorization, they're proof to identity. If a system wants to use identity information to extrapolate authorization levels, that isn't OAuth based authorization. It's a custom mode of authorization. Not wrong, it's just not OAuth based authorization using access tokens.

Reply:
@missBerg Thanks for pointing this out! I think claims from both the Access Token and ID token should be able to be used for authorization. We need to figure out a way to express both the `contians` matching and the `exact` match against an element in an array.